play-services-fido/core/src/main/kotlin/org/microg/gms/fido/core/transport/TransportHandler.kt

```diff
@@ -50,21 +70,31 @@ abstract class TransportHandler(val transport: Transport, val callback: Transpor
     private suspend fun ctap2register(
         connection: CtapConnection,
         options: RequestOptions,
-        clientDataHash: ByteArray
+        clientDataHash: ByteArray,
+        pinToken: ByteArray? = null
     ): Pair<AuthenticatorMakeCredentialResponse, ByteArray?> {
         connection.capabilities
         val reqOptions = AuthenticatorMakeCredentialRequest.Companion.Options(
             when (options.registerOptions.authenticatorSelection?.residentKeyRequirement) {
                 RESIDENT_KEY_REQUIRED -> true
                 RESIDENT_KEY_PREFERRED -> connection.hasResidentKey
                 RESIDENT_KEY_DISCOURAGED -> false
-                else -> options.registerOptions.authenticatorSelection?.requireResidentKey == true
+                else -> when(options.registerOptions.authenticatorSelection?.requireResidentKey) {
+                    false -> false
+                    else -> connection.hasResidentKey
+                }
             },
+
             when (options.registerOptions.authenticatorSelection?.requireUserVerification) {
                 REQUIRED -> true
                 DISCOURAGED -> false
                 else -> connection.hasUserVerificationSupport
             }
+            // According to newer drafts of CTAP2.1, the user verification key MUST NOT be included
+            // if the authenticator is not capable of built-in verification
```

Comment:
It is stated in the v2.2 review draft: https://fidoalliance.org/specs/fido-v2.2-rd-20230321/fido-client-to-authenticator-protocol-v2.2-rd-20230321.html#authenticatorGetAssertion. However, since that is still a review draft, and it deprecates the use of the "uv" flag and mandates the use of a pinUvAuthParam option (which I have not implemented), I have dropped this from the code

Reply:
I see. In this case we should probably just bail out early if `requireUserVerification` is set to `required` and the authenticator doesn't support it. That way we still honor the requirement from WebAuthN correctly.

Reply:
I have written it so that if no PIN is provided, the requireUserVerification variable is passed to the authenticator in the "uv" flag. On my Yubikey, this leads to a status code of 0x2b (unsupported error). I haven't changed the code that deals with CTAP error messages, so this is currently logged as an exception and then the program continues.